### ISSUE TITLE
[RFE] Add a progress/download indicator

### ISSUE THREAD

**Issue by neitsab** [nspawn/nspawn#19]
Hey, many thanks for your work on this project! I's really helpful and improves systemd-nspawn usability quite a bunch :-)

I'd like to submit a feature request for adding some information regarding the fact a download is occurring whenever one runs `nspawn -i <image>` with a new image.

Currently on the very first run, after importing the key it hangs blankly while the download is occurring. Even knowing what was happening, I still had to check the system monitor to verify my bandwidth was being used and that it didn't just freeze.

The connection I'm using being capped at 1.5 MBps certainly doesn't help making it snappy, but I do think this is the kind of situation that would deserve printing some messages about the success of the key import and/or that it is currently downloading an image (depending on the situation).

Here's an example first run:

```
% sudo !!
sudo nspawn -i archlinux/archlinux/tar
/etc/systemd/import-pubring.gpg does not exist
Do you want to create it [y/n]: y
gpg: le trousseau local « /etc/systemd/import-pubring.gpg » a été créé
gpg: /root/.gnupg/trustdb.gpg : base de confiance créée
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2304  100  2304    0     0   3250      0 --:--:-- --:--:-- --:--:--  3249
gpg: clef 0E2A1E4B25A425F6 : clef publique « Nspawn.org Team (Nspawn.org Signing key for Images and Containers) <team@nspawn.org> » importée
gpg:       Quantité totale traitée : 1
gpg:                     importées : 1

[here it hanged for several minutes without printing anything]

Removing read-only flag if present from image archlinux-archlinux-tar...
Image deployed locally as archlinux-archlinux-tar. Details:

Name=archlinux-archlinux-tar
Path=/var/lib/machines/archlinux-archlinux-tar
Type=directory
ReadOnly=no
CreationTimestamp=Fri 2024-09-06 19:23:59 CEST
Usage=18446744073709551615
Limit=infinity
UsageExclusive=18446744073709551615
LimitExclusive=infinity
```

Thanks!

**Comment by neitsab** [comment]
I just noticed that there was actually a progress indicator but directly sent to the journal:

```journal
2024-09-06T19:20:52+02:00 zenbook systemd-importd[272046]: (transfer1) Pulling 'https://hub.nspawn.org/storage/archlinux/archlinux/tar/image.tar.xz', saving as 'archlinux-archlinux-tar'.
2024-09-06T19:20:52+02:00 zenbook systemd-importd[272046]: (transfer1) Operating on image directory '/var/lib/machines'.
2024-09-06T19:20:52+02:00 zenbook systemd-importd[272046]: (transfer1) Downloading 566B for https://hub.nspawn.org/storage/archlinux/archlinux/tar/SHA256SUMS.gpg.
2024-09-06T19:20:52+02:00 zenbook systemd-importd[272046]: (transfer1) Acquired 566B.
2024-09-06T19:20:52+02:00 zenbook systemd-importd[272046]: (transfer1) Download of https://hub.nspawn.org/storage/archlinux/archlinux/tar/SHA256SUMS.gpg complete.
2024-09-06T19:20:52+02:00 zenbook systemd-importd[272046]: (transfer1) HTTP request to https://hub.nspawn.org/storage/archlinux/archlinux/tar/image.nspawn failed with code 404.
2024-09-06T19:20:52+02:00 zenbook systemd-importd[272046]: (transfer1) Settings file could not be retrieved, proceeding without.
2024-09-06T19:20:52+02:00 zenbook systemd-importd[272046]: (transfer1) Downloading 166.2M for https://hub.nspawn.org/storage/archlinux/archlinux/tar/image.tar.xz.
2024-09-06T19:20:53+02:00 zenbook systemd-importd[272046]: (transfer1) Downloading 79B for https://hub.nspawn.org/storage/archlinux/archlinux/tar/SHA256SUMS.
2024-09-06T19:20:53+02:00 zenbook systemd-importd[272046]: (transfer1) Acquired 79B.
2024-09-06T19:20:53+02:00 zenbook systemd-importd[272046]: (transfer1) Download of https://hub.nspawn.org/storage/archlinux/archlinux/tar/SHA256SUMS complete.
2024-09-06T19:20:55+02:00 zenbook systemd-importd[272046]: (transfer1) Got 1% of https://hub.nspawn.org/storage/archlinux/archlinux/tar/image.tar.xz. 5min 31s left at 508.5K/s.
2024-09-06T19:20:57+02:00 zenbook systemd-importd[272046]: (transfer1) Got 2% of https://hub.nspawn.org/storage/archlinux/archlinux/tar/image.tar.xz. 4min 34s left at 608K/s.
2024-09-06T19:20:59+02:00 zenbook systemd-importd[272046]: (transfer1) Got 3% of https://hub.nspawn.org/storage/archlinux/archlinux/tar/image.tar.xz. 4min 9s left at 660.8K/s.
...
2024-09-06T19:23:59+02:00 zenbook systemd-importd[272046]: (transfer1) Acquired 626.4M.
2024-09-06T19:23:59+02:00 zenbook systemd-importd[272046]: (transfer1) Download of https://hub.nspawn.org/storage/archlinux/archlinux/tar/image.tar.xz complete.
2024-09-06T19:23:59+02:00 zenbook systemd-importd[272046]: (transfer1) SHA256 checksum of https://hub.nspawn.org/storage/archlinux/archlinux/tar/image.tar.xz is valid.
2024-09-06T19:23:59+02:00 zenbook systemd-importd[272046]: (transfer1) gpg: Signature faite le dim. 02 juin 2024 03:19:52 CEST
2024-09-06T19:23:59+02:00 zenbook systemd-importd[272046]: (transfer1) gpg:                avec la clef RSA 9E31BD4963FC2D19815FA7180E2A1E4B25A425F6
2024-09-06T19:23:59+02:00 zenbook systemd-importd[272046]: (transfer1) gpg: Bonne signature de « Nspawn.org Team (Nspawn.org Signing key for Images and Containers) <team@nspawn.org> » [inco>
2024-09-06T19:23:59+02:00 zenbook systemd-importd[272046]: (transfer1) gpg: Attention : utilisation d'une clef sans confiance.
2024-09-06T19:23:59+02:00 zenbook systemd-importd[272046]: (transfer1) Signature verification succeeded.
2024-09-06T19:24:00+02:00 zenbook systemd-importd[272046]: (transfer1) Created new local image 'archlinux-archlinux-tar'.
2024-09-06T19:24:00+02:00 zenbook systemd-importd[272046]: (transfer1) Operation completed successfully.
2024-09-06T19:24:00+02:00 zenbook systemd-importd[272046]: (transfer1) Exiting.
```

So maybe simply redirect it to stdout?

**Comment by Edu4rdSHL** [comment]
Hello, @neitsab. Thank you for your suggestion. It's done purposely actually via https://github.com/nspawn/nspawn/blob/09c522033f85ae3f602d3c5f3e6ecf3cb3d679b2/nspawn#L144 (see the output redirection), you can remove it if you want to see the progress. I'm not adding more effort to the current bash script really but on https://github.com/nspawn/nspawn/tree/rust_dev which has command autocompletion, better design and many more things.

**Comment by neitsab** [comment]
Hey @Edu4rdSHL, thanks for the quick reply and for pointing out the line responsible for the redirection, I'll do that on my local version to know what's happening :-)

 Great to know you are rewriting the program in Rust, I hadn't seen the branch. I'll close this issue then :-)

**Comment by Edu4rdSHL** [comment]
I will reopen it just to keep track of this (I think it's disabled on the Rust version too).

**Comment by neitsab** [comment]
Sorry, I realized you might want to do that since you had added the "good first issue" label, but only after closing 😅

Is the Rust port already usable? If you need beta testers I wouldn't mind giving it a try. However right now I'm facing another issue when trying to launch the container with `machinectl start`:

```
2024-09-06T19:39:45+02:00 zenbook systemd[1]: Virtual Machine and Container Storage (Compatibility) was skipped because of an unmet condition check (ConditionPathExists=/var/lib/machines.ra>
2024-09-06T19:39:45+02:00 zenbook systemd[1]: Starting Load Kernel Module dm_mod...
2024-09-06T19:39:45+02:00 zenbook systemd[1]: Starting Load Kernel Module loop...
2024-09-06T19:39:45+02:00 zenbook systemd[1]: Starting Load Kernel Module tun...
2024-09-06T19:39:45+02:00 zenbook systemd[1]: modprobe@dm_mod.service: Deactivated successfully.
2024-09-06T19:39:45+02:00 zenbook systemd[1]: Finished Load Kernel Module dm_mod.
2024-09-06T19:39:45+02:00 zenbook systemd[1]: modprobe@loop.service: Deactivated successfully.
2024-09-06T19:39:45+02:00 zenbook systemd[1]: Finished Load Kernel Module loop.
2024-09-06T19:39:45+02:00 zenbook systemd[1]: modprobe@tun.service: Deactivated successfully.
2024-09-06T19:39:45+02:00 zenbook systemd[1]: Finished Load Kernel Module tun.
2024-09-06T19:39:45+02:00 zenbook systemd[1]: Starting Container archlinux-archlinux-tar...
2024-09-06T19:39:45+02:00 zenbook systemd-nspawn[273312]: Using default interface naming scheme 'v255'.
2024-09-06T19:39:45+02:00 zenbook systemd-nspawn[273312]: Network interface name 've-archlinux-archlinux-tar' has been changed to 've-archlinu891E' to fit length constraints.
2024-09-06T19:39:45+02:00 zenbook systemd-nspawn[273312]: Failed to add new veth interfaces (ve-archlinu891E:host0): Operation not supported
2024-09-06T19:39:45+02:00 zenbook systemd-nspawn[273317]: Parent died too early
2024-09-06T19:39:45+02:00 zenbook systemd[1]: run-systemd-nspawn-unix\x2dexport-archlinux\x2darchlinux\x2dtar.mount: Deactivated successfully.
2024-09-06T19:39:45+02:00 zenbook systemd[1]: systemd-nspawn@archlinux-archlinux-tar.service: Main process exited, code=exited, status=1/FAILURE
2024-09-06T19:39:45+02:00 zenbook systemd[1]: systemd-nspawn@archlinux-archlinux-tar.service: Failed with result 'exit-code'.
2024-09-06T19:39:45+02:00 zenbook systemd[1]: Failed to start Container archlinux-archlinux-tar.
```

Should I open another issue, or is that linked to https://blog.nspawn.org/posts/wireless-adapters-on-systemd-nspawn-containers/ and I just need to configure my Wifi adapter in an `.nspawn` file?

**Comment by neitsab** [comment]
~~Nevermind, seems I hit https://bugzilla.redhat.com/show_bug.cgi?id=2001631 (which I found via https://docs.fedoraproject.org/en-US/fedora-server/containerization/systemd-nspawn-setup/#_resolution_for_physical_interfaces_managed_by_networkmanager). I'll work it out.~~

Edit: Scratch that, I had updated my system (including the kernel) but hadn't rebooted yet. After a reboot `sudo machinectl start archlinux-archlinux-tar` works, although now I have to find the login credentials 😁

--> done in https://github.com/nspawn/mkosi-definitions/blob/master/hub-images/ArchLinux/archlinux/mkosi.images/00-archlinux-tar.conf#L36 (`root`/`root`). Is that documented somewhere and I missed it?

**Comment by Edu4rdSHL** [comment]
> Is the Rust port already usable? If you need beta testers I wouldn't mind giving it a try. 

Yes, it's usable and has even more features than the script at this point.

> Edit: Scratch that, I had updated my system (including the kernel) but hadn't rebooted yet. After a reboot sudo machinectl start archlinux-archlinux-tar works, although now I have to find the login credentials 😁

https://nspawn.org/images/ - All images have the login: root/root

**Comment by neitsab** [comment]
> Yes, it's usable and has even more features than the script at this point.

Great, I'll give it a try then! I'm not super familiar with building Rust apps, is a simple `cargo build` in the project dir enough? If I get it to work I can package it in the AUR too.

> https://nspawn.org/images/ - All images have the login: root/root

🙏🏻

**Comment by neitsab** [comment]
The current version from the rust_dev branch actually displays a download progress indicator and I'm loving it! 😍

**Comment by Edu4rdSHL** [comment]
> The current version from the rust_dev branch actually displays a download progress indicator and I'm loving it! 😍

Awesome! It was quite a long time ago when I wrote it, so I didn't remember exactly, but glad to hear that it's already there.

**Comment by Edu4rdSHL** [comment]
Implemented in https://github.com/nspawn/nspawn/releases/tag/1.1.1
